src/app/modules/form/admin-form/admin-form.controller.ts

```diff
@@ -1642,6 +1642,113 @@ export const _handleCreateFormField: RequestHandler<
+ * @returns 403 when user does not have permissions to create logic
+ * @returns 404 when form cannot be found
+ * @returns 422 when user in session cannot be retrieved from the database
+ * @returns 500 when database error occurs
+ */
+export const _handleCreateLogic: RequestHandler<
+  { formId: string },
+  LogicDto | ErrorDto,
+  LogicDto
+> = (req, res) => {
+  const { formId } = req.params
+  const createdLogic = req.body
+  const sessionUserId = (req.session as Express.AuthedSession).user._id
+
+  // Step 1: Retrieve currently logged in user.
+  return (
+    UserService.getPopulatedUserById(sessionUserId)
+      .andThen((user) =>
+        // Step 2: Retrieve form with write permission check.
+        AuthService.getFormAfterPermissionChecks({
+          user,
+          formId,
+          level: PermissionLevel.Write,
+        }),
+      )
+      // Step 3: Create form logic
+      .andThen((retrievedForm) =>
+        AdminFormService.createFormLogic(retrievedForm, createdLogic),
+      )
+      .map((createdLogic) => res.status(StatusCodes.OK).json(createdLogic))
+      .mapErr((error) => {
+        logger.error({
+          message: 'Error occurred when creating form logic',
+          meta: {
+            action: 'handleCreateLogic',
+            ...createReqMeta(req),
+            userId: sessionUserId,
+            formId,
+            createdLogic,
+          },
+          error,
+        })
+        const { errorMessage, statusCode } = mapRouteError(error)
+        return res.status(statusCode).json({ message: errorMessage })
+      })
+  )
+}
+
+/**
+ * Shape of request body used for joi validation for create and update logic
+ */
+const joiLogicBody = {
+  logicType: Joi.string()
+    .valid(...Object.values(LogicType))
+    .required(),
+  conditions: Joi.array()
+    .items(
+      Joi.object({
+        field: Joi.string().required(),
+        state: Joi.string()
+          .valid(...Object.values(LogicConditionState))
+          .required(),
+        value: Joi.alternatives()
+          .try(
+            Joi.number(),
+            Joi.string(),
+            Joi.array().items(Joi.string()),
+            Joi.array().items(Joi.number()),
+          )
+          .required(),
+        ifValueType: Joi.string()
+          .valid(...Object.values(LogicIfValue))
+          .required(),
+      }).unknown(true),
+    )
+    .required(),
+  show: Joi.alternatives().conditional('logicType', {
+    is: LogicType.ShowFields,
+    then: Joi.array().items(Joi.string()).required(),
+  }),
+  preventSubmitMessage: Joi.alternatives().conditional('logicType', {
+    is: LogicType.PreventSubmit,
+    then: Joi.string().required(),
+  }),
+}
+
+/**
+ * Handler for POST /forms/:formId/logic
+ */
+export const handleCreateLogic = [
+  celebrate({
+    [Segments.BODY]: Joi.object({
+      ...joiLogicBody,
+      // Allow other field related key-values to be provided and let the model
```

Comment:
Since this is a new endpoint, we can (and should) strictly control the validation. Suggest removing `.unknown(true)`, then it can just be

```ts
[Segments.BODY]: joiLogicBody
```

Reply:
edited